src/context/map.tsx

```diff
@@ -82,64 +102,70 @@ export const MapContextProvider = ({
+            if (layer.id === currentLayer.id) {
+              return { ...layer, isActive };
+            }
+            return layer;
+          }) ?? [],
+      );
+
+      if (currentLayer.geojson) return;
+
       const geojson =
         currentLayer.geojson ?? (await getGeoJSON(currentLayer.geojsonUrl));
-      setLayers(prevLayers => {
-        if (prevLayers === null) {
-          return prevLayers;
-        }
 
-        return prevLayers.reduce((list: Layer[], layer) => {
-          const { defaultActive, ...nextLayer } = layer;
-          if (nextLayer.id !== id) {
-            list.push(nextLayer);
-          } else {
-            list.push({
-              ...nextLayer,
-              isActive,
-              geojson: geojson as FeatureCollection,
-            });
-          }
-          return list;
-        }, []);
-      });
+      setLayers(
+        prevLayers =>
+          prevLayers?.map(layer => {
+            if (layer.id === currentLayer.id) {
+              return { ...layer, geojson: geojson as FeatureCollection };
+            }
+            return layer;
+          }) ?? [],
+      );
     },
     [getLayerById],
   );
 
+  // When active layers are changed, fetch the geojson
   useEffect(() => {
-    if (
-      layersFromParams === '' &&
-      layers?.some(item => 'defaultActive' in item)
-    ) {
-      const layersID = layers
-        .filter(item => item.defaultActive)
-        .map(({ id }) => id);
-      const nextLayerSearchParams = getUrlSearchParamsForLayers(
-        layersFromParams,
-        layersID,
-        true,
-      );
-      const text = params.get('text');
-      router.replace(
-        `${pathname}${nextLayerSearchParams}${
-          text ? `&text=${encodeURIComponent(text)}` : ''
-        }`,
-      );
-    }
-  }, [layers, layersFromParams, params, pathname, router]);
+    layers
+      ?.filter(({ isActive }) => isActive)
+      .forEach(async currentLayer => {
+        if (!currentLayer.geojson) {
+          const geojson = await getGeoJSON(currentLayer.geojsonUrl);
+          setLayers(
+            prevLayers =>
+              prevLayers?.map(layer => {
+                if (layer.id === currentLayer.id) {
+                  return { ...layer, geojson: geojson as FeatureCollection };
+                }
+                return layer;
+              }) ?? [],
+          );
+        }
+      });
+  }, [layers]);
 
+  // Update URL when active layers are changed
   useEffect(() => {
-    const [activatedLayers, disabledLayers] = partition(layers ?? [], item =>
-      layersFromParams.split(',').map(Number).includes(item.id),
+    const textFromParams = params.get('text')
+      ? `&text=${params.get('text')}`
+      : '';
+    const nextLayerSearchParams = `?layers=${layers
+      ?.filter(({ isActive }) => isActive)
+      .map(({ id }) => id)
+      .join(',')}`;
+    window.history.pushState(
+      null,
+      '',
+      `${nextLayerSearchParams}${textFromParams}`,
```

Comment:
Why not `router.push` from `useRouter`?

Reply:
In this case, `router.push` kept reloading the page and seding a few requests, by using `history` (since we're in a client component) the URL is updated without any impact to the app